avr-hal-generic/src/i2c.rs

```diff
@@ -425,51 +514,57 @@ macro_rules! impl_twi_i2c {
+                //    to the TWI Address Register (TWAR), TWI Status Register (TWSR), and TWI Data
+                //    Register (TWDR) must be complete before clearing this flag
                 //
-                // NOTE: In the start() fn above sets the `twint` flag, which `I think` is setting it to true:
+                // NOTE: the start() fn above sets the `twint` flag, setting it to `true`:
                 //    
                 //      .$twint().set_bit()
                 //
-                // But 'true' === 1 which in the language of the spec here means `cleared`...
+                // But in rust, a bool 'true' === 1. It's confusing that the datasheet says a bit
+                // is cleared by writing a logic one to it:
                 //
                 //    The TWINT Flag must be cleared by software by writing a logic one to it.
                 //
-                // Here in avr-device/src/generic.rs
-                //
-                //    impl<FI> FieldReader<bool, FI> {
-                //        /// Value of the field as raw bits.
-                //        #[inline(always)]
-                //        pub fn bit(&self) -> bool {
-                //            self.bits
-                //        }
-                //        /// Returns `true` if the bit is clear (0).
-                //        #[inline(always)]
-                //        pub fn bit_is_clear(&self) -> bool {
-                //            !self.bit()
-                //        }
-                //        /// Returns `true` if the bit is set (1).
-                //        #[inline(always)]
-                //        pub fn bit_is_set(&self) -> bool {
-                //            self.bit()
-                //        }
-                //    }
+                // And the FieldReader also explicitly uses 'clear' to refer to 0 and 'set' to
+                // refer to 1:
                 //
-                while self.p.$twcr.read().$twint().bit_is_clear() { }
-            }
+                //    impl<FI> FieldReader<bool, FI> { /// Value of the field as raw bits.
+                //    #[inline(always)] pub fn bit(&self) -> bool { self.bits } /// Returns `true`
+                //    if the bit is clear (0).  #[inline(always)] pub fn bit_is_clear(&self) ->
+                //    bool { !self.bit() } /// Returns `true` if the bit is set (1).
+                //    #[inline(always)] pub fn bit_is_set(&self) -> bool { self.bit() } }
+                //
+                // QUESTION: so is what I'm seeing a distinction between a _HARDWARE FLAG_ being 'cleared',
+                // (perhaps 'reset' is a better name?) vs. setting/clearing a _bit_ in a bitfield?
+                while self.p.$twcr.read().$twint().bit_is_clear() { } }
 
             fn transact(&mut self) {
+                // QUESTION: why are we setting TWEN here? It has already been set (above). Does it
```

Comment:
We set `$twen` with every other bit write. I suppose it doesn't hurt, but I'm curious why it is necessary. Is this to guard against downstream code setting this flag and breaking this library?

Reply:
Simplifying a bit, you can only ever write a whole register, in this case, all 8 bits at once.  For toggling only certain bits, you'd have to first read the entire register, modify its bits with some logic and and or operations and then write the modified value back.  This is also called a _read-modify-write_ loop.

The `svd2rust` API specifies that a `.write(...)` will write the _whole_ register.  This means all 8 bits are written.  Any bit that is not explicitly specified will be set to its "reset"-value which is usually just 0.

Thus, if we were to not set `TWEN` to 1 every time, we'd actually be turning off the TWI peripheral.

---

Now, you could of course use `.modify(...)` which is `svd2rust`'s version of the _read-modify-write_ loop.  Then we'd only need to care about the bits we're actually interacting with.

This does, however, have a number of disadvantages:

1. Its at least 3 more instructions per write, which can amount to quite a large overhead in the big picture.
2. Because we leave some bits untouched, it can quickly lead to hard-to-discover bugs because some bit was still set where it shouldn't.  By keeping it explicit which bits are set and which are cleared for every write, it is much easier to see such circumstances.
3. For a single 8-bit register, the `.write()` operation is atomic.  A _read-modify-write_ loop however is not, which means it might need a critical section to prevent an interrupt firing in the middle.

Reply:
This is like an intro course into micro-controllers. Thx!